File: Problem067-Add Binary/Solution.cpp
/*
 * @SEU(master);@ZZULI(bachelor)
 * author:CuiXuan
 * email:873059043@qq.com
 * If you find any errors, please let me know.
 * If you have any better solution, please let me know.
 *
 * Solution Using time 4ms
 */
class Solution {
public:
    string addBinary(string a, string b) {
        int len1=a.length();
        int len2=b.length();
        int len = (len1>len2?len1:len2)+5;
        int i=len1-1,j=len2-1,top=0;
        char* str = (char*)malloc(sizeof(char)*len);
        memset(str,0,sizeof(char)*len);
        while(true){
            if(i<0 && j<0)break;
            else if(i<0)str[top++]+=b[j--]-'0';
            else if(j<0)str[top++]+=a[i--]-'0';
            else str[top++]+=a[i--]-'0'+b[j--]-'0';
        }
        for(i=0;i<len;i++){
            str[i+1]+=str[i]/2;
            str[i]%=2;
            str[i]+='0';
        }
        i--;
        while(i>0 && str[i]=='0')i--;
        j=i;str[i+1]=0;
        i=0;
        while(i<j){
            char tmp = str[i];
            str[i]=str[j];
            str[j]=tmp;
            i++;j--;
        }
        return string(str);
    }
};

File: Problem067-Add Binary/Solution.py
"""
 @SEU(master);@ZZULI(bachelor)
 author:CuiXuan
 email:873059043@qq.com
 If you find any errors, please let me know.
 If you have any better solution, please let me know.

 Solution Using time 76ms
"""
class Solution(object):
    def addBinary(self, a, b):
        """
        :type a: str
        :type b: str
        :rtype: str
        """
        len1=len(a);len2=len(b);top=0
        lens = len1+2 if len1>len2 else len2+2
        str = [0 for i in range(lens)]
        i=len1-1;j=len2-1
        while True:
            if i<0 and j<0:break
            elif i<0:str[top]+=ord(b[j])-ord('0');top+=1;j-=1
            elif j<0:str[top]+=ord(a[i])-ord('0');top+=1;i-=1
            else:str[top]+=ord(b[j])-ord('0')+ord(a[i])-ord('0');top+=1;i-=1;j-=1
        i=0
        while i<lens-1:
            str[i+1]+=str[i]/2
            str[i]%=2
            i+=1
        while i>0 and str[i]==0:i-=1
        ret = []
        for i in range(i+1):ret.append(chr(str[i]+ord('0')))
        j=i;i=0;
        while i<j:
            tmp = ret[i]
            ret[i]=ret[j]
            ret[j]=tmp
            i+=1;j-=1
        return "".join(ret)